### PR TITLE
[Ripple] Removing default values when states are set

### PR DESCRIPTION
**Pull request by galiak11**
### Description
When components are setting default values for states it changes the basic rule that all states should inherit from the `normal` state, unless customized by the user.
Currently, the default value for the `normal` and `selected` states are blue. If a client app is setting the `normal` state to red, then the `selected` state is blue, instead of red.

The fix keeps the current default values, but removes them as soon as any state is customized by [setRippleColor:forState:].

### Issue
b/130344265

### CODE REVIEW

File: components/Ripple/src/MaterialState.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+ The integert value used for each State matches the UIControlState value so both states
+ maybe used interchangeably. For example:
+
+ [rippleView setRippleColor:UIColor.redColor forState:(MaterialState)UIControlStateHighlighted];
+
+ @note Additional Material states that do not have a corresponding UIControlState value are
+      within UIControlState's range that is reserved for application use.
+ */
+typedef NS_OPTIONS(NSInteger, MaterialState) {
```
Comment (romoore) [inline components/Ripple/src/MaterialState.h]: The PR summary states that this can be used in APIs that accept UIControlState values. Does this work for Swift as well? 

If we are introducing new values for UIControlState API arguments, I think it's better for us to actually provide UIControlState values.  That would mean also providing an appropriate bitmask and performing any recommended internal announcements/discussions about our intention to take control of those values within Google.  I think we should be *very cautious* about introducing these values and allowing their use on UIControl subclasses (like MDCButton or MDCChip). 

My biggest concern is that a client may be using the same Application-reserved range for other purposes on an MDC component and that our reassignment of the values can cause unexpected side-effects.  I think it's important to note that `UIControlStateApplication` is named *Application* and not *User* or something else. Given the number of levels of downstream frameworks that consume our components before they get to an application, there's a great opportunity for conflicts.

**Review by romoore — CHANGES_REQUESTED** [review]
Please provide an internal team design doc + review of this proposal. I think we should also perform internal (to Google) announcements and a thorough review of client code to ensure we're not overlapping.

File: components/Ripple/src/MaterialState.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  /** Used as Material Selected state. Matches UIControlStateSelected value. */
+  MaterialStateSelected = 1 << 2,
+
+  /** Used as Material Active state. Matches UIControlState's reserved range for applications. */
+  MaterialStateActive = 1 << 20,
+  /** Used as Material Dragged state. Matches UIControlState's reserved range for applications. */
+  MaterialStateDragged = 1 << 21,
+  /** Used as Material Error state. Matches UIControlState's reserved range for applications. */
+  MaterialStateError = 1 << 22,
+};
```
Comment (jverkoey) [inline components/Ripple/src/MaterialState.h]: Please follow our naming conventions for APIs by using the MDC prefix for all symbols here and throughout.

File: components/Ripple/src/MaterialState.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
```
Comment (jverkoey) [inline components/Ripple/src/MaterialState.h]: This file looks like something that should live in its own directory as a standalone component.

File: components/Ripple/src/MDCStatefulRippleView.h [hunk 4]
```diff
@@ -14,22 +14,6 @@
...
- - MDCRippleStateHighlighted: The ripple is activated and shown.
- - MDCRippleStateSelected: The ripple is in the selected state.
- - MDCRippleStateDragged: The ripple is in the dragged state.
- */
-typedef NS_OPTIONS(NSInteger, MDCRippleState) {
-  MDCRippleStateNormal = 0,
-  MDCRippleStateHighlighted = 1 << 0,
-  MDCRippleStateSelected = 1 << 1,
-  MDCRippleStateDragged = 1 << 2,
-};
```
Comment (jverkoey) [inline components/Ripple/src/MDCStatefulRippleView.h]: This is a breaking change for clients; what would a more gradual migration plan look like for clients?

File: components/Ripple/src/MaterialState.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+typedef NS_OPTIONS(NSInteger, MaterialState) {
+
+  /** Used as Material Enabled or Resting state. Matches UIControlStateNormal value. */
+  MaterialStateNormal = 0,
+  /** Used as Material Pressed state. Matches UIControlStateHighlighted value. */
+  MaterialStateHighlighted = 1 << 0,
+  /** Used as Material Disabled state. Matches UIControlStateDisabled value. */
+  MaterialStateDisabled = 1 << 1,
+  /** Used as Material Selected state. Matches UIControlStateSelected value. */
+  MaterialStateSelected = 1 << 2,
```
Comment (jverkoey) [inline components/Ripple/src/MaterialState.h]: Consider literally equating these values to UIControlState equivalents.

File: components/Ripple/tests/unit/MaterialStateTests.m [hunk 6]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
```
Comment (jverkoey) [inline components/Ripple/tests/unit/MaterialStateTests.m]: These tests should likely also live in a separate directory.

**Review by jverkoey — CHANGES_REQUESTED** [review]
This looks like a good direction overall. For historical purposes, please add a go/ link to the design doc in the PR description that documents:

1. This pattern.
2. The expected work required to roll this out to our components.
3. The rejected alternatives.

File: components/Ripple/src/MaterialState.h (hunk 1 above)
Comment (jverkoey) [reply components/Ripple/src/MaterialState.h]: Agreed with the sentiment here and the encouragement of caution.

I believe the risk of overlapping is somewhat mitigated because we own the components for which these new control states apply. I'd expect overlap to exist if someone were to subclass one of our components and add new states, but I expect this is a rarity.

Comment (jverkoey) [reply components/Ripple/src/MaterialState.h]: Spoke with Rob offline a bit more about his concerns, summarizing them here for visibility.

In essence this PR proposal would reserve certain bits of UIControlState for MDC-specific use. This will not affect all UIControlState use in an app, but it could potentially impact anyone that is subclassing our components and adding new UIControlState values. As it stands, many of our stateful components can be subclassed, so it will be healthy for us to reach out to the community to evaluate the impact of reserving some UIControlState bits at the MDC level. Notably, I'm hopeful the community will share any subclasses that may have reserved their own UIControlState bits so that we can define a migration path or influence which bits we decide to reserve for MDC.

In essence, we'll want to start a request for comment for the following proposal: "we'd like to reserve the following UIControlState bits (defined bits here) for all MDC classes; this will affect you if you subclass any of these components and add your own states, and we'd like to better understand the impact of this proposal"

**Comment by galiak11** [comment]
Removing state enumeration. This PR now only fixes the default values for ripple.

**Comment by yarneo** [comment]
I would suggest opening a new PR @galiak11 as it looks like the PR has an entirely different intent than the original one.

**Comment by romoore** [comment]
@galiak11 Please also link to the Issue for this PR. The b/ link you provided is for a design document, not changing the defaults on Ripple (or a bug caused by the current behavior).

File: components/Ripple/src/MDCStatefulRippleView.h [hunk 7]
```diff
@@ -17,7 +17,6 @@
 /**
  Provides the current state of the ripple. The ripple is either in its normal state, or in the
  selected state where the ripple remains spread on the view.
-
```
Comment (romoore) [inline components/Ripple/src/MDCStatefulRippleView.h:20]: Please revert this change - it doesn't seem to be necessary.

File: components/Ripple/src/MDCStatefulRippleView.m [hunk 8]
```diff
@@ -54,37 +55,45 @@ - (instancetype)initWithCoder:(NSCoder *)coder {
 }
 
 - (void)commonMDCStatefulRippleViewInit {
-  if (_rippleColors == nil) {
```
Comment (romoore) [inline components/Ripple/src/MDCStatefulRippleView.m:57]: Why does this `nil` check need to change? I would expect both arrays to be `nil` during initialization. If not, please add a separate check for `_rippleColors`.

File: components/Ripple/src/MDCStatefulRippleView.m [hunk 9]
```diff
@@ -54,37 +55,45 @@ - (instancetype)initWithCoder:(NSCoder *)coder {
...
         [UIColor colorWithWhite:0 alpha:kDefaultRippleDraggedAlpha];
-    _rippleColors[@(MDCRippleStateSelected | MDCRippleStateDragged)] =
+    _rippleDefaultColors[@(MDCRippleStateSelected | MDCRippleStateDragged)] =
         [selectionColor colorWithAlphaComponent:kDefaultRippleDraggedAlpha];
   }
 }
 
 - (UIColor *)rippleColorForState:(MDCRippleState)state {
-  UIColor *rippleColor = _rippleColors[@(state)];
+  UIColor *rippleColor = _rippleColors[@(state)] ?: _rippleDefaultColors[@(state)];
```
Comment (romoore) [inline components/Ripple/src/MDCStatefulRippleView.m:81]: This logic is very complex and behaves very differently from UIControl (the only stateful API I'm really familiar with). Could you please add comments/docs to the method to help outline the updated logic?

File: components/Ripple/src/MDCStatefulRippleView.m [hunk 10]
```diff
@@ -99,6 +108,7 @@ - (void)updateActiveRippleColor {
 
 - (void)setRippleColor:(UIColor *)rippleColor forState:(MDCRippleState)state {
   _rippleColors[@(state)] = rippleColor;
+  [_rippleDefaultColors removeAllObjects];
```
Comment (romoore) [inline components/Ripple/src/MDCStatefulRippleView.m:111]: So if I understand this logic correctly, once I override the value for `.normal` state, I lose all the defaults for the `.selected` and `.dragged` states?   This seems not great, since I'd expect my changes to only affect (1) the `.normal` state or (2) any state that was previously "falling back" to `.normal`.

File: components/Ripple/src/MDCStatefulRippleView.m [hunk 11]
```diff
@@ -171,7 +181,7 @@ - (void)setRippleHighlighted:(BOOL)rippleHighlighted {
     // If ripple becomes highlighted we initiate a ripple with animation.
     [self updateRippleColor];
     [self beginRippleTouchDownAtPoint:_lastTouch animated:_didReceiveTouch completion:nil];
-  } else {
+  } else if (!rippleHighlighted) {
```
Comment (romoore) [inline components/Ripple/src/MDCStatefulRippleView.m:184]: Is this necessary for this change? It seems like an independent bug that could be fixed in a separate PR.

**Review by romoore — CHANGES_REQUESTED** [review]
The changed behavior seems unintuitive with a lot of unexpected side-effects. I think if this logic is absolutely necessary, a migration flag should be used so clients can opt-in to the "remove all" behavior.

**Comment by galiak11** [comment]
Duplicate of #7337
